Add ESM samples troubleshooting guide (#459)
Esri/jsapi-resources: TROUBLESHOOTING.md

**Step 3 - Reproduce.** Try to reproduce the issue in a simple, focused stand-alone HTML/JavaScript app. This will help determine if the issue is related to the Maps SDK, or not. Sometimes, in the process of trying to reproduce the error you will solve the issue.

**Step 4 – Create a sample.** Always try to provide a simple sample when posting your question to the Esri Community Site. There great online options for sharing samples including [codepen](https://codepen.io/), [github.com](https://github.com/), [codesandbox](https://codesandbox.io/) or [StackBlitz](https://stackblitz.com/) are great for providing reproducible examples. Code snippets often do not provide enough information about a project because they don’t include information on configuration, dependencies or coding patterns. Providing a working sample is essential to allow others to quickly reproduce your issue. Sometimes the process of creating the sample helps solve the problem.

## Understand the cause of an issue

> With all the various parts in modern build systems, determining where the problem is occurring is critical. There are typically three categories where problems occur: configuration, bundler/framework, or the Maps SDK.

Reviewer: ```suggestion
With all the various parts in modern build systems, determining where the problem is occurring is critical. There are typically three categories where problems occur: configuration, bundler/framework, or the JS Maps SDK.
```

> ## Configuration issues

Reviewer 2: Empty newline after header items, couple cases here. Do newlines matter if header 2 or 3?

Author: Ah yes, I can remove them. Github markdown ignores them, but I'll fix them for consistency.

> ### Configuration issues

Reviewer 3: ```suggestion
## Configuration issues
```

The most common errors are related to project configuration. Most of these types of issues look like bugs, but oftentimes they are not:
*	Module bundler – e.g. webpack and rollup configuration
*	Framework – e.g. Angular, React configuration

> *	Package.json – e.g. dependency version mismatches

Reviewer 3: ```suggestion
*	package.json – e.g. dependency version mismatches
```

It's odd seeing that capitalized.

*	Maps SDK – e.g. esriConfig, OAuth tokens, API keys

#### Example Configuration issues

**Problem 1:** `404` errors that indicate a file is missing in a production build, but dev build works fine.

_Potential Resolution:_ Check the bundler or framework documentation on setting the base path for production builds. See [example](https://github.com/Esri/jsapi-resources/issues/436).  

**Problem 2:** Module parse failed: Unexpected token. You may need an appropriate loader to handle this file type. 

_Potential Resolution:_ These are typically related to transpiling. You may be missing a needed transpiler plugin or need to upgrade to a newer version of the module bundler. The error messages may point to one or more Maps SDK modules. See [example](https://github.com/Esri/jsapi-resources/issues/424).
*	Get a screenshot of the full error message

> *	Narrow down an Maps SDK version where the break occurred

Reviewer: ```suggestion
*	Narrow down a JS Maps SDK version where the break occurred
```

**Problem 3:** Error loading a .wasm or .woff file.

_Potential Resolution:_ Configure your web server with the correct MIME types. See [example](https://developers.arcgis.com/javascript/latest/install-and-set-up/#web-server-hosting-configuration).

> ### Bundler/Framework issues

Reviewer 3: ```suggestion
## Bundler/Framework issues
```

Bundlers and frameworks do occasionally have their own bugs. If you are working with a major framework or bundler, they have very active communities. Searching for the error message using a general internet search is typically the fastest way to get information and hints.
* Configuration issues
* Component issues
* Syntax errors
* Life-cycle problems

#### Example Bundler/Framework issues

> We do come across bugs in module bundlers and frameworks that are unrelated to the Maps SDK. Sometimes these bugs have already been documented. Sometimes, we recommend you open an issue on the framework or bundler repository, that’s where many examples of these issues can be found, for example:

Reviewer: ```suggestion
We do come across bugs in module bundlers and frameworks that are unrelated to the JS Maps SDK. Sometimes these bugs have already been documented. Sometimes, we recommend you open an issue on the framework or bundler repository, that’s where many examples of these issues can be found, for example:
```

*	React issues: https://github.com/facebook/react/issues 
*	Angular issues: https://github.com/angular/angular-cli/issues 
*	Rollup issues: https://github.com/rollup/rollup/issues 

> *	Webpack issues: https://github.com/webpack/webpack/issues 

Reviewer 3: Sort?

Author: Done.

> ### Maps SDK issues

Reviewer 3: ```suggestion
## Maps SDK issues
```

Errors occurring in the Maps SDK typically have one of these types of errors:
*	Errors that occur at runtime and not during the build process
*	Errors that occur in a Maps SDK module
*	Errors that can be reproduced using the Maps SDK in plain JavaScript without a framework

#### Example Maps SDK issue

**Errors in production builds.** While very rare, sometimes the production build process of a framework creates errors that don’t show up in developer builds. These are often related to bundling, minification or transpiling. Look for hints in the error message that might point to functionality inside a Maps SDK module.

[Example - ReferenceError: Cannot access “e” before initialization](https://github.com/Esri/jsapi-resources/issues/309). In this case, when the error stack dump was expanded in the browser console, you could see file names that appear to be from the Maps SDK such as _renderLegendForLayer_ and _renderLegendForElement_, these are excellent hints.

> Verified errors in production builds can be reported on the [Maps SDK Community site](https://community.esri.com/t5/arcgis-javascript-maps-sdk-questions/bd-p/arcgis-api-for-javascript-questions), or if it’s specifically related to an ES modules sample on this repo, then you can open a github issue [here](https://github.com/Esri/jsapi-resources/issues).

Reviewer 3: ```suggestion
Verified errors in production builds can be reported on the [Maps SDK Community site](https://community.esri.com/t5/arcgis-javascript-maps-sdk-questions/bd-p/arcgis-api-for-javascript-questions), or if it’s specifically related to an ES modules sample on this repo, then you can open a GitHub issue [here](https://github.com/Esri/jsapi-resources/issues).
```

**Regressions.** This is when Maps SDK functionality works in one release, and then the same functionality is broken in the next one. It’s important to insure the bugs are related to some other dependency in package.json. When testing for regressions, you will change only the version of @arcgis/core and keep all other dependencies constant. 

Most Maps SDK regressions can be re-created in a plain JavaScript application, without the framework or bundler. The ESM CDN is an easy way to test and prototype these types of issues: https://github.com/Esri/jsapi-resources/tree/master/esm-samples/jsapi-esm-cdn. 

Verified regressions can be reported to[ Esri Technical Support](https://support.esri.com/en/contact-tech-support) or the [Maps SDK Community site](https://community.esri.com/t5/arcgis-javascript-maps-sdk-questions/bd-p/arcgis-api-for-javascript-questions).

> **New bugs.** The best practice for verifying these is to re-create them in a simple, plain JavaScript application, without the framework or bundler. The ESM CDN is an easy way to test for these types of issues, here is sample application. 

Reviewer 3: Period after "issues"? Link to sample?

Author: Fixed.
